Holovibes/cuda_sources/bw_area_filter.cu

```diff
@@ -0,0 +1,354 @@
+    int x = (idx / width) * 2 + 1;
+    int y = (idx % width) * 2 + 1;
+    idx = x * width + y;
+
+    if (y >= 1 && x >= 1 && x < (width - 1) && y < (height - 1) && !IS_BACKGROUND(image_d[idx]))
+    {
+        size_t neighbors[4];
+        size_t nb_neighbors = 0;
+
+        for (int k = -1; k <= 1; k += 2)
+        {
+            size_t jdx = x * width + y + k;
+            if (labels_d[jdx])
+                neighbors[nb_neighbors++] = jdx;
+            jdx = (x + k) * width + y;
+            if (labels_d[jdx])
+                neighbors[nb_neighbors++] = jdx;
+        }
+
+        check_and_update_link(labels_d, linked_d, labels_sizes_d, idx, neighbors, nb_neighbors, mutex);
+    }
+}
+
+void first_pass(const float* image_d,
+                size_t* labels_d,
+                size_t* linked_d,
+                size_t* labels_sizes_d,
+                size_t* size_t_gpu_,
+                const size_t width,
+                const size_t height,
+                const cudaStream_t stream)
+{
+    // int* mutex;
+    // cudaXMalloc(&mutex, sizeof(int));
+    cudaXMemset(size_t_gpu_, 0, sizeof(int));
+
+    size_t size = width * height;
+    uint threads = get_max_threads_1d();
+    uint blocks = map_blocks_to_problem(size / 2, threads);
+    cudaXMemset(linked_d, 0, sizeof(size_t));
+
+    first_pass_kernel1<<<blocks, threads, 0, stream>>>(image_d, labels_d, linked_d, labels_sizes_d, width, height);
+    cudaDeviceSynchronize();
+    first_pass_kernel2<<<blocks, threads, 0, stream>>>(image_d,
+                                                       labels_d,
+                                                       linked_d,
+                                                       labels_sizes_d,
+                                                       width,
+                                                       height,
+                                                       size_t_gpu_);
+    cudaDeviceSynchronize();
+    first_pass_kernel3<<<blocks, threads, 0, stream>>>(image_d,
+                                                       labels_d,
+                                                       linked_d,
+                                                       labels_sizes_d,
+                                                       width,
+                                                       height,
+                                                       size_t_gpu_);
+    cudaDeviceSynchronize();
+    first_pass_kernel4<<<blocks, threads, 0, stream>>>(image_d,
+                                                       labels_d,
+                                                       linked_d,
+                                                       labels_sizes_d,
+                                                       width,
+                                                       height,
+                                                       size_t_gpu_);
+    cudaDeviceSynchronize();
```

Comment:
No need to use cudaDeviceSynchronize when following calculation are done on GPU. It should be used when you need result on CPU side